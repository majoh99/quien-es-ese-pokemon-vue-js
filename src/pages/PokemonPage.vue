<template>
  <div class="page-container d-flex">
    <div class="col pokemon-img d-flex items-center justify-center">
      <h2 v-if="!pokemon">Espere, por favor...</h2>
      <div v-else class="d-flex items-center justify-center picture-container">
        <img class="logo-mobile" src="@/assets/quien-es-ese-pokemon.webp" alt="¿Quién es ese Pokemon?">
        <PokemonPicture 
          :pokemon-id="pokemon.id" 
          :show-pokemon="showPokemon" 
        />
      </div>
      
    </div>
    <div v-if="pokemon" class="col pokemon-options bg-red">
      <img class="logo" src="@/assets/quien-es-ese-pokemon.webp" alt="¿Quién es ese Pokemon?">
      <PokemonOptions 
        :pokemons="pokemonArr" 
        @selection-pokemon="checkAnswer" 
        :status-option="status"
        :pokemon-id="pokemon.id"
        :is-correct="isCorrect"
       />
      <template v-if="showAnswer">
        <p class="fade-in text-white">{{ message }}</p>
        <div class="d-flex justify-center">
          <button class="btn-newGame d-flex items-center text-white" @click="newGame">
            <img src="@/assets/icons/play_circle.svg" alt="play_circle">
            Volver a jugar
          </button>
        </div>
      </template>
    </div>
  </div>
  <!-- <h2 v-if="!pokemon">Espere por favor...</h2>
  <div v-else>
    <img class="logo" src="@/assets/quien-es-ese-pokemon.webp" alt="¿Quién es ese Pokemon?">
    <PokemonPicture :pokemon-id="pokemon.id" :show-pokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonArr" @selection-pokemon="checkAnswer" :status-option="status" />
    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button class="btn-newGame" @click="newGame">Nuevo Juego</button>
    </template>
  </div> -->
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'

import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
  name: 'PokemonPage',
  components: { PokemonPicture, PokemonOptions },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: '',
      status: false,
      isCorrect: null,
    }
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions()
      const rndInt = Math.floor( Math.random() * 4 )
      this.pokemon = this.pokemonArr[ rndInt ]
    },
    checkAnswer( selectedId ) {
      this.showPokemon = true
      this.showAnswer = true
      this.pokemon.id === selectedId
        ? (this.isCorrect = true)
        : (this.isCorrect = false)

      if (selectedId === this.pokemon.id) {
        const name = this.pokemon.name
        this.message = `Correcto, es ${ name.charAt(0).toUpperCase() + name.slice(1) }`
        this.status = true

      } else {
        const name = this.pokemon.name
        this.message = `Oops, era ${ name.charAt(0).toUpperCase() + name.slice(1) }`
        this.status = true
      }
    },
    newGame() {
      this.showPokemon = false,
      this.showAnswer = false,
      this.pokemonArr = [],
      this.pokemon = null,
      this.status = false,
      this.isCorrect = null;
      this.mixPokemonArray()
    }
  },
  mounted() {
    this.mixPokemonArray()
  }
}
</script>

<style scoped>
.bg-red{
  background-color: var(--background-color);
}

.btn-newGame{
  background-color: transparent;
  border-radius: 5px;
  border: none;
  cursor: pointer;
  font-size: 16px;
  margin-bottom: 10px;
  padding: 12px 6px;
}
</style>

