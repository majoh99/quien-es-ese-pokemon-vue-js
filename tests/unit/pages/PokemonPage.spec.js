import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage Component', () => {
    
    let wrapper

    beforeEach(() => {
        wrapper = shallowMount( PokemonPage )
    })

    test('debe de hacer match con el snapshot', () => {        
        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('debe de llamar mixPokemonArray al montar', () => {
        
        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' )
        const wrapper = shallowMount( PokemonPage )
        
        expect( mixPokemonArraySpy ).toHaveBeenCalled()
    })

    test('debe de hacer match con el snapshot cuando cargan los pokemons', () => {
        
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect( wrapper.html() ).toMatchSnapshot()

    })
    
    
    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {

        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        expect( picture.exists() ).toBeTruthy()
        expect( options.exists() ).toBeTruthy()

        
        expect( picture.attributes('pokemonid') ).toBe('5')
        expect( options.attributes('pokemons') ).toBeTruthy()

        
    })
    

    test('pruebas con checkAnswer', async() => {
        
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(5)

        expect( wrapper.find('p').exists() ).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBe(true)
        expect( wrapper.find('p').text() ).toBe(`Correcto, es ${ pokemons[0].name.charAt(0).toUpperCase() + pokemons[0].name.slice(1) }`)

        await wrapper.vm.checkAnswer(10)

        expect( wrapper.vm.message ).toBe(`Oops, era ${ pokemons[0].name.charAt(0).toUpperCase() + pokemons[0].name.slice(1) }`)

    })
    

})