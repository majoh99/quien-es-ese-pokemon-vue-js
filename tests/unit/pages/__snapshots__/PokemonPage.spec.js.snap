// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot 1`] = `
<div class="page-container d-flex">
  <div class="col pokemon-img d-flex items-center justify-center">
    <h2>Espere, por favor...</h2>
  </div>
  <!--v-if-->
</div>
<!-- <h2 v-if="!pokemon">Espere por favor...</h2>
  <div v-else>
    <img class="logo" src="@/assets/quien-es-ese-pokemon.webp" alt="¿Quién es ese Pokemon?">
    <PokemonPicture :pokemon-id="pokemon.id" :show-pokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonArr" @selection-pokemon="checkAnswer" :status-option="status" />
    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button class="btn-newGame" @click="newGame">Nuevo Juego</button>
    </template>
  </div> -->
`;

exports[`PokemonPage Component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div class="page-container d-flex">
  <div class="col pokemon-img d-flex items-center justify-center">
    <div class="d-flex items-center justify-center picture-container"><img class="logo-mobile" src="" alt="¿Quién es ese Pokemon?">
      <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
    </div>
  </div>
  <div class="col pokemon-options bg-red"><img class="logo" src="" alt="¿Quién es ese Pokemon?">
    <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]" statusoption="false" pokemonid="5"></pokemon-options-stub>
    <!--v-if-->
  </div>
</div>
<!-- <h2 v-if="!pokemon">Espere por favor...</h2>
  <div v-else>
    <img class="logo" src="@/assets/quien-es-ese-pokemon.webp" alt="¿Quién es ese Pokemon?">
    <PokemonPicture :pokemon-id="pokemon.id" :show-pokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonArr" @selection-pokemon="checkAnswer" :status-option="status" />
    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button class="btn-newGame" @click="newGame">Nuevo Juego</button>
    </template>
  </div> -->
`;
